### ISSUE TITLE
Item Migration: update item status assignment

### ISSUE THREAD

**Issue by wwelling** [TAMULib/mod-data-migration#222]
The discrepancy is:  the numeric value from the case statement in the query shouldn't be used as an index to assign anything. That value exists only to create the order by statement. The folio status name is assigned by taking the $item_status_object{'name'} value from the item_status_object that is in position 0 in the item_status_array. And that name value is derived by comparing the voyager item_status to the %vger_status_folio_status hash.
If you read through the comments in the perl script starting at line 113 I hope it will all become clear. I admit that I didn't use the best naming conventions throughout; it would have been clearer if I had.

```
227     item_status=14  vger_status_folio_status{hr->{'item_status'}}==Aged to lost
229     vger_status_name: Lost--System Applied  item_status_codes{14}: Lost--System Applied
227     item_status=1   vger_status_folio_status{hr->{'item_status'}}==Available
229     vger_status_name: Not Charged   item_status_codes{1}: Not Charged
235     $VAR1 = [
          {
            'vger_status_name' => 'Lost--System Applied',
            'date' => '2018-05-16T09:31:06Z',
            'vger_status' => '14',
            'name' => 'Aged to lost'
          },
          {
            'name' => 'Available',
            'vger_status' => '1',
            'date' => '2019-01-15T09:30:49Z',
            'vger_status_name' => 'Not Charged'
          }
        ];
256     folio_status: Aged to lost
276     i == 0, therefore folio_status(Aged to lost) is assigned as {'status'}{'name'}
256     folio_status: Available
$VAR1 = {
          '4533522' => {
                         'mfhd' => '5907794',
                         'folio_item' => {
                                           'status' => {
                                                         'date' => '2018-05-16T09:31:06Z',
                                                         'name' => 'Aged to lost'
                                                       },
                                           'discoverySuppress' => 'true',
                                           'statisticalCodeIds' => [
                                                                     '38c86b2b-8156-4b7e-943e-460e15ea0dc0',
                                                                     '37e178fa-dd83-441d-bb2f-77de049cfba1',
                                                                     '71b2451d-408f-4ac7-ad28-e5e774efd06a'
                                                                   ]
                                         },
                         'mfhdloc' => '238'
                       }
        };
```

**Comment by wwelling** [comment]
extract_item_voyager_sub.pl
```
#!/m1/shared/bin/perl
#####################################################################################################################################
# Developed by Anne L. Highsmith, Texas A&M University Libraries, 20190110
#####################################################################################################################################
# Designed to create folio item records in json, based on mfhds listed in data/holdings-storage/holdings directory
#####################################################################################################################################
#	20200130 changed processing for voyager and folio item statuses. As of Q4, folio item status became an enumerated list.
#	Folio enumerated list will be expanded later; as of this date handling item status as follows: 
#		1) mapping voyager item status to folio item status only for those statuses where available folio status makes sense
#		2) copying all voyager statuses into statistical codes.
#	Undoubtedly all of this will change as 3-pronged item state is implemented
# 20200604 Made some modifications to how item status is handled on the assumption TAMU would be using circulation BL modules to do
#	loan migration
#####################################################################################################################################
use strict;
use warnings;
use JSON;
use lib qw(/export/home/voyager/perl/lib);
use Data::Dumper;
use URI::Escape;
use LWP::UserAgent;
use UCLA_Batch;
use MARC::Batch;
use TAMU_DBConnect;
use TAMU_FOLIO;
use TAMU_General;
my $config = shift;
my $config_ref = &read_config($config);
my $dbase = shift;
my $err;
my $errfile = "/export/home/voyager/folio/preview/data/item-storage/items/${dbase}_extract_errors.txt";
open(ERR, ">$errfile") or die "Cannot open $errfile\n";

my $mode = 'test';
my $host = ($mode eq 'prod') ? "obrian" : "surprise";
my $dbh = &oracle_connect("$host", "$dbase", 'read');
my $status_date_utc = q/TO_CHAR(cast(item_status_date as timestamp) at time zone 'UTC', 'YYYY-MM-DD"T"HH24:MI:SS"Z"')/;
my %items;
my %item_status_codes = (
    1 => 'Not Charged',
    2 => 'Charged',
    3 => 'Renewed',
    4 => 'Overdue',
    5 => 'Recall Request',
    6 => 'Hold Request',
    7 => 'On Hold',
    8 => 'In Transit',
    9 => 'In Transit Discharged',
   10 => 'In Transit On Hold',
   11 => 'Discharged',
   12 => 'Missing',
   13 => 'Lost--Library Applied',
   14 => 'Lost--System Applied',
   15 => 'Claims Returned',
   16 => 'Damaged',
   17 => 'Withdrawn',
   18 => 'At Bindery',
   19 => 'Cataloging Review',
   20 => 'Circulation Review',
   21 => 'Scheduled',
   22 => 'In Process',
   23 => 'Call Slip Request',
   24 => 'Short Loan Request',
   25 => 'Remote Storage Request'
);
#############################################################################################
## 20200512 Realized, while working with loan migration, that I can't set Charged, Renewed,
## overdue, etc. to the 'Checked out' in FOLIO because theb I can't migrate the loan using
## the technique I'm using. Changed any such status to 'Available' on assumption loan
## migration would catch it and update it.
## 20210115 Changed 7, Awaiting pickup to available for same reason. May also need to change
##          all of the In transit statuses also.
#############################################################################################
my %vger_status_folio_status = ( 
    1 => 'Available',            
    2 => 'Available',          
    3 => 'Available',          
    4 => 'Available',          
    7 => 'Available',      
    8 => 'In transit',           
    9 => 'In transit',           
   10 => 'In transit',           
   11 => 'Available',            
   12 => 'Missing',              
   13 => 'Declared lost',        
   14 => 'Aged to lost',
   15 => 'Claimed returned',
   17 => 'Withdrawn',
   22 => 'In process',           
);

my %custodian_codes = (
	'amdb' => '38c86b2b-8156-4b7e-943e-460e15ea0dc0',
	'msdb' => 'f5d1069d-f718-4f0e-8ff8-308f55540daf'
);

MAINPROG: {
#	#my $mfhd_to_folio_id_ref = &extract_mfhd_ids($dbase);
#	#my $items_ref = &extract_item_ids($mfhd_to_folio_id_ref);
#	#my $mfhd_to_folio_item_id_ref = &read_mfhd_item_links($dbase);
	my $location_file_ref = &read_locations;
	my $material_types_ref = &expand_material_types($config);
	my $stat_code_ref = &expand_stat_codes($config);
#
#	#&load_base_item($items_ref,$mfhd_to_folio_id_ref, $mfhd_to_folio_item_id_ref, $location_file_ref, $stat_code_ref);
	my %items;
	$items{'4533522'}{'mfhd'} = '5907794';
	$items{'4533522'}{'mfhdloc'} = '238';
	my $items_ref = \%items;
	&set_item_status($items_ref, $location_file_ref,$stat_code_ref);
	print Dumper $items_ref;
}
#####################################################################################################################################
# ITEM STATUS PROCESSING 
#	
#	For each item
#		While there are item statuses left to process for the item
#			select a number corresponding to the priority of the item status code, and the item status date
#			store the item status number in the item status object as the 'vger_status'
#			store the item status code in the item status object as the 'vger_status_name'
#			using the vger_status_folio_status hash, store the folio status name in the item_status_object as 'name'
#     store the item status date in the item status object as 'date'
#			push the item status object onto the item status array
#
#		At the end of the while loop, the item status array will have an item status object for each
#			item_status that was assigned to the item in the item_status_table. Items can have an indefinite number of statuses 
#			in Voyager. The entries in the item status array will be sorted in the order of priority of the item_status
#			as determined by TAMU circ staff.
#
#		For each object in the item status array
#			Extract the name of the item status from the item status object, assigning 'Available' as a default if no name appears
#			Use the vger_status_name as an index to assign the appropriate statistical code unless the vger_status_name indicates
#				the item is damaged, which is a separate data element in folio
#			Set the suppressDiscovery flag = 'true' if the status indicates the item is missing or no longer held
#			If the index variable indicates this is the first object in the item status array (position 0), then it is the
#				most important status by priority and the 'name' and 'date' elements should be assigned to the folio item as folio status and date.
#
#	At the end of the subroutine, the folio item record will have an item status, which is a required element, and 1 or more
#		statistical codes, which correspond to all of the item statuses the item had in Voyager at the point of migration, using the
#		original voyager item status codes
# 20200604
#		If the circtrans element is defined, indicating the item appears in the circ_transactions table, set folio item status to
#			'Available' so that loan can be processed by BL modules during loan migration
#####################################################################################################################################

sub set_item_status {
	my ($items_ref, $location_file_ref,$stat_code_ref) = @_;
	my $sql = "
	select	item_status, $status_date_utc AS item_status_date, ct.circ_transaction_id AS circtrans,
       case 
      	     when item_status_desc ='Charged'                         then '01'
      	     when item_status_desc ='Renewed'                         then '02'
      	     when item_status_desc ='Overdue'                         then '03'
      	     when item_status_desc ='Recall Request'                  then '04'
      	     when item_status_desc ='Hold Request'                    then '05'
      	     when item_status_desc ='On Hold'                         then '06'
      	     when item_status_desc ='Missing'                         then '07'
      	     when item_status_desc ='Lost--Library Applied'           then '08'      	     
      	     when item_status_desc ='Lost--System Applied'            then '09'
      	     when item_status_desc ='In Transit'                      then '10'
      	     when item_status_desc ='In Transit Discharged'           then '11'
      	     when item_status_desc ='In Transit On Hold'              then '12'
      	     when item_status_desc ='Withdrawn'                       then '13'      
      	     when item_status_desc ='Claims Returned'                 then '16'
      	     when item_status_desc ='Damaged'                         then '17'        
      	     when item_status_desc ='At Bindery'                      then '18'     
      	     when item_status_desc ='Cataloging Review'               then '19'
      	     when item_status_desc ='Circulation Review'              then '20'
      	     when item_status_desc ='Scheduled'                       then '21'      
      	     when item_status_desc ='In Process'                      then '22'     
      	     when item_status_desc ='Call Slip Request'               then '23'
      	     when item_status_desc ='Short Loan Request'              then '24'
      	     when item_status_desc ='Remote Storage Request'          then '25'
	      	end revised_status
	from  	$dbase.item_status istat,
	      	$dbase.item_status_type itype,
	      	$dbase.circ_transactions ct
	where 	istat.item_id = ?
	and   	istat.item_status = itype.item_status_type
	and   	istat.item_id = ct.item_id(+)
	order by 
	      case 
      	     when item_status_desc ='Charged'                         then '01'
      	     when item_status_desc ='Renewed'                         then '02'
      	     when item_status_desc ='Overdue'                         then '03'
      	     when item_status_desc ='Recall Request'                  then '04'
      	     when item_status_desc ='Hold Request'                    then '05'
      	     when item_status_desc ='On Hold'                         then '06'
      	     when item_status_desc ='Missing'                         then '07'
      	     when item_status_desc ='Lost--Library Applied'           then '08'      	     
      	     when item_status_desc ='Lost--System Applied'            then '09'
      	     when item_status_desc ='In Transit'                      then '10'
      	     when item_status_desc ='In Transit Discharged'           then '11'
      	     when item_status_desc ='In Transit On Hold'              then '12'
      	     when item_status_desc ='Withdrawn'                       then '13'      
      	     when item_status_desc ='Claims Returned'                 then '16'
      	     when item_status_desc ='Damaged'                         then '17'        
      	     when item_status_desc ='At Bindery'                      then '18'     
      	     when item_status_desc ='Cataloging Review'               then '19'
      	     when item_status_desc ='Circulation Review'              then '20'
      	     when item_status_desc ='Scheduled'                       then '21'      
      	     when item_status_desc ='In Process'                      then '22'     
      	     when item_status_desc ='Call Slip Request'               then '23'
      	     when item_status_desc ='Short Loan Request'              then '24'
      	     when item_status_desc ='Remote Storage Request'          then '25'
	      	end
	";
	my $sth = 	$dbh->prepare ("$sql") || die $dbh->errstr;

	foreach my $item_id ( sort bynum keys %{ $items_ref}) {
		my @item_status_array;
		$sth->bind_param(1,$item_id);
		$sth->execute || die $dbh->errstr;
		
		while (my $hr = $sth->fetchrow_hashref('NAME_lc')) {
			my %item_status_object;
			$item_status_object{'vger_status'} = $hr->{'item_status'};
			if (defined $hr->{'circtrans'}) {
				$item_status_object{'name'} = 'Available';
			}
			elsif (defined $hr->{'item_status'}) {
				$item_status_object{'name'} = $vger_status_folio_status{$hr->{'item_status'}};
			}
			else {
				$item_status_object{'name'} = 'Available';
			}
			print __LINE__, "\titem_status=$hr->{'item_status'}\tvger_status_folio_status{hr->{'item_status'}}==$vger_status_folio_status{$hr->{'item_status'}}\n";
			$item_status_object{'vger_status_name'} = $item_status_codes{$hr->{'item_status'}};
			print __LINE__, "\t", "vger_status_name: $item_status_object{'vger_status_name'}\titem_status_codes{$hr->{'item_status'}}: $item_status_codes{$hr->{'item_status'}}\n";
			if (defined $hr->{'item_status_date'} ) {
				$item_status_object{'date'} = $hr->{'item_status_date'};
			}
			push(@item_status_array,\%item_status_object);
		}
		print __LINE__, "\t", Dumper \@item_status_array;
		#####################################################################
		# Assume before starting the loop that discoverySuppress should be
		# false, i.e. that item should be displayed in discovery system
		# If any of its statuses indicates it's no longer in the collection,
		#	however, the discoverySuppress will be set to true
		#####################################################################
		$items_ref->{$item_id}{'folio_item'}{'discoverySuppress'} = 'false';
		my $custodian_stat_code_id = $custodian_codes{$dbase};
		push (@{ $items_ref->{$item_id}{'folio_item'}{'statisticalCodeIds'} }, $custodian_stat_code_id);

		for (my $i=0; $i <= $#item_status_array; $i++){
			my $folio_status;
			if (defined $item_status_array[$i]{'name'}) {
				$folio_status = $item_status_array[$i]{'name'};
			}
			# using 'Available' as a bogus status for those which don't yet exist
			#	in FOLIO
			else {
				$folio_status = 'Available';
			}
			print __LINE__, "\tfolio_status: $folio_status\n";
			my $stat_code_name = $item_status_array[$i]{'vger_status_name'};
			my $status_date = $item_status_array[$i]{'date'};
			if ($stat_code_name !~ /damaged/i) {
				my $stat_code_id = $stat_code_ref->{$stat_code_name}{'id'};
				push (@{ $items_ref->{$item_id}{'folio_item'}{'statisticalCodeIds'} }, $stat_code_id);
			}
			else {
				$items_ref->{$item_id}{'folio_item'}{'itemDamagedStatusId'} = '54d1dd76-ea33-4bcb-955b-6b29df4f7930';
			}
			###################################################################
			# 1st if sequence resets discoverySuppress if needed
			###################################################################			
			if ($stat_code_name =~ /(Missing|Lost|Withdrawn)/) {
				$items_ref->{$item_id}{'folio_item'}{'discoverySuppress'} = 'true';
			}
			###################################################################
			# Now, set item status for most import status available
			###################################################################						
			if ($i == 0) {
				print __LINE__, "\ti == 0, therefore folio_status($folio_status) is assigned as {'status'}{'name'}\n";
				$items_ref->{$item_id}{'folio_item'}{'status'}{'name'} = $folio_status;
				$items_ref->{$item_id}{'folio_item'}{'status'}{'date'} = $status_date;
			}
		}
	}
}
######################################################################################################################################
sub bynum {$a <=> $b}
```

**Comment by wwelling** [comment]
Correction required in item status map.
